hazelcast/src/main/java/com/hazelcast/client/impl/management/ClientConnectionProcessListener.java

```diff
@@ -0,0 +1,212 @@
+        childListeners.forEach(listener -> listener.possibleAddressesCollected(addresses));
+    }
+
+    @Override
+    public void authenticationSuccess(Address address) {
+        childListeners.forEach(listener -> listener.authenticationSuccess(address));
+    }
+
+    @Override
+    public void credentialsFailed(Address address) {
+        childListeners.forEach(listener -> listener.credentialsFailed(address));
+    }
+
+    @Override
+    public void clientNotAllowedInCluster(Address address) {
+        childListeners.forEach(listener -> listener.clientNotAllowedInCluster(address));
+    }
+
+    @Override
+    public void clusterConnectionFailed(String clusterName) {
+        childListeners.forEach(listener -> listener.clusterConnectionFailed(clusterName));
+    }
+
+    @Override
+    public void clusterConnectionSucceeded(String clusterName) {
+        childListeners.forEach(listener -> listener.clusterConnectionSucceeded(clusterName));
+    }
+
+    @Override
+    public void remoteClosedConnection(Address address) {
+        childListeners.forEach(listener -> listener.remoteClosedConnection(address));
+    }
+
+    @Override
+    public ClientConnectionProcessListener withAdditionalListener(ClientConnectionProcessListener listener) {
+        childListeners.add(listener);
+        return this;
+    }
+}
+
+/**
+ * Implementations can be attached to a {@link com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager} to receive
+ * fine-grained events about the client-to-cluster connection process.
+ *
+ * @since 5.2
+ */
+public interface ClientConnectionProcessListener
+        extends EventListener {
+
+    ClientConnectionProcessListener NOOP = new ClientConnectionProcessListener() {
+
+        @Override
+        public ClientConnectionProcessListener withAdditionalListener(ClientConnectionProcessListener listener) {
+            return listener;
+        }
+    };
+
+    /**
+     * Triggered before trying to connect to an address. This event is triggered after a {@link #possibleAddressesCollected(List)}
+     * event, and the {@code address} parameter is always an element of the possible-addresses list (passed to
+     * {@link #possibleAddressesCollected(List)}).
+     * <p>
+     * If the connection fails, this event is followed by either a {@link #remoteClosedConnection(Address)} or a
+     * {@link #connectionAttemptFailed(Address)} event.
+     * <p>
+     * If the connection is established but the client runs into an authentication failure, then this will be indicated by a
+     * subsequent {@link #clientNotAllowedInCluster(Address)} or a {@link #credentialsFailed(Address)} event.
+     * <p>
+     * If the authentication succeeds then a {@link #authenticationSuccess(Address)} event will be fired, followed by an
+     * {@link #clusterConnectionSucceeded(String)}.
+     *
+     * @param address
+     */
+    default void attemptingToConnectToAddress(Address address) {
```

Comment:
Will these be delivered on the connection thread? Will they possibly slow down the connection process? Or what if they throw exception?

Reply:
Yes, on the connection thread. The [implementations are very quick jobs](https://github.com/hazelcast/management-center/pull/5799/files#diff-e61f21cfa563f131fc719f6ed7de6f0467be7b3c893fbe1dae24571311f70b7bR47) so they are unlikely to slow down anything. 

I added per-listener exception handling in my most recent commit.